### PR TITLE
Specify forbidden fields if metadata properties validation failed

### LINKED ISSUE #100933: CRD structural schema validation doesn't allow having description on openAPIV3Schema.properties[metadata]
#### What happened:
A CRD with the following schema
```
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: foos.example.com
spec:
  group: example.com
  names:
    kind: Foo
    listKind: FooList
    plural: foos
    singular: foo
  scope: Cluster
  versions:
  - name: v1beta1
    served: true
    storage: true
    schema:
      openAPIV3Schema:
        description: A foo.
        type: object
        properties:
          metadata:
            type: object
            description: Additional validation on foo's name.         <--  this is the violator
            properties:
              name:
                description: The name may only contain lowercase letters.
                type: string
                pattern: '^[a-z]+$'
```
will fail server side validation
```
$ kubectl apply -f crd3.yaml
The CustomResourceDefinition "foos.example.com" is invalid: spec.validation.openAPIV3Schema.properties[metadata]: Forbidden: must not specify anything other than name and generateName, but metadata is implicitly specified
```

while removing the description on metadata will make the CRD valid
```
$ cat crd3.yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: foos.example.com
spec:
  group: example.com
  names:
    kind: Foo
    listKind: FooList
    plural: foos
    singular: foo
  scope: Cluster
  versions:
  - name: v1beta1
    served: true
    storage: true
    schema:
      openAPIV3Schema:
        description: A foo.
        type: object
        properties:
          metadata:
            type: object
            properties:
              name:
                description: The name may only contain lowercase letters.
                type: string
                pattern: '^[a-z]+$'
```
```
$ kubectl apply -f crd3.yaml
customresourcedefinition.apiextensions.k8s.io/foos.example.com created
```

#### What you expected to happen:
Either we allow description, or we make the error message clear. 

#### How to reproduce it (as minimally and precisely as possible):
Create the example CRD above.  

#### Anything else we need to know?:
The validation code is [here](https://github.com/kubernetes/kubernetes/blob/669016067d49110a35769f7db42705cc5f0becff/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go#L162). If we want to allow setting description, I think we should clear the description field the same way as we do for `metadata.Type` before calling `reflect.DeepEqual`.

/good-first-issue
/area custom-resources
/sig api-machinery

#### Environment:
- Kubernetes version (use `kubectl version`): HEAD

### PR DESCRIPTION
**Pull request by ilyee**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Add more error messages for metadata properties validation.

When some forbidden properties(for ex, `description`) are specified in CRD yaml, error message will be "must not specify any property other than name and generateName: description" rather than "must not specify anything other than name and generateName, but metadata is implicitly specified", which will make errors message less confusing.

#### Which issue(s) this PR fixes:

Fixes #100933 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go [hunk 1]
```diff
@@ -152,21 +153,27 @@ func validateStructuralInvariants(s *Structural, lvl level, fldPath *field.Path)
...
 		metadata.Type = ""
+		metadata.Properties = nil
 		metadata.Default.Object = nil // this is checked in API validation (and also tested)
 		if metadata.ValueValidation == nil {
 			metadata.ValueValidation = &ValueValidation{}
 		}
 		if !reflect.DeepEqual(metadata, Structural{ValueValidation: &ValueValidation{}}) {
 			// TODO: this is actually a field.Invalid error, but we cannot do JSON serialization of metadata here to get a proper message
-			allErrs = append(allErrs, field.Forbidden(fldPath.Child("properties").Key("metadata"), "must not specify anything other than name and generateName, but metadata is implicitly specified"))
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("properties").Key("metadata"), "must not specify anything other than type and properties, but metadata is implicitly specified"))
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go:176]: this needs tests before this (potentially functionally) change to prove that this PR does not break the existing API.

**Comment by sttts** [comment]
Better error messages a great. I miss test coverage though to be confident this PR does not break the existing API, i.e. does not change the behaviour, but only the error message. Please add those before you do the change of the error message logic.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go (hunk 1 above)
Comment (ilyee) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go:176]: done

**Comment by ilyee** [comment]
Unit tests have been added, PTAL if free @sttts

**Comment by sttts** [comment]
Can you split the unit tests from the commit, run them before your changes (even in separate PR), and then add your changes on-top? Background: we must be 100% sure behaviour has not changed.

**Comment by ilyee** [comment]
> Can you split the unit tests from the commit, run them before your changes (even in separate PR), and then add your changes on-top? Background: we must be 100% sure behaviour has not changed.

Agree with it, i'll separate test and changes.

**Comment by ilyee** [comment]
> Can you split the unit tests from the commit, run them before your changes (even in separate PR), and then add your changes on-top? Background: we must be 100% sure behaviour has not changed.

unit tests have been separated into #103402, PTAL if you have time, thx @sttts

**Comment by dims** [comment]
Is this PR still needed, please rebase if so (or we can close it?)

**Comment by ilyee** [comment]
> Is this PR still needed, please rebase if so (or we can close it?)

I'll rebase it soon

**Comment by ilyee** [comment]
@sttts sorry for the late update, since unit tests have been merged, PTAL at this pr if free, thx!

**Comment by ilyee** [comment]
@sttts some bugs are fixed, pls review it, thx

**Comment by pacoxu** [comment]
https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=102596
image

/hold
the integration failed consistently. Would you fix the CI and then `/unhold`?

**Pull request closed without merging**